File: backend/app/agents/summarizer.py
"""
AI agent for generating podcast summaries.
Creates concise 200-300 word summaries of podcast content.
"""

import re
from typing import Dict, Any

from app.agents.base_agent import BaseAgent, AgentProcessingError
from app.config import get_settings
from app.utils.logger import get_logger

logger = get_logger(__name__)
settings = get_settings()


class SummarizerAgent(BaseAgent):
    """
    AI agent that generates concise summaries of podcast transcripts.
    
    Creates professional, informative summaries between 200-300 words
    that capture the main topics, key points, and overall theme of the podcast.
    """

    def process(self, transcript_content: str, **kwargs) -> Dict[str, Any]:
        """
        Generate a summary of the podcast transcript.
        
        Args:
            transcript_content: The raw transcript text to summarize
            **kwargs: Additional parameters (unused)
            
        Returns:
            Dictionary with 'summary' key containing the generated summary
            
        Raises:
            AgentProcessingError: If summarization fails
        """
        logger.info(f"[{self.agent_name}] Starting summarization of transcript",
                   transcript_length=len(transcript_content),
                   word_count=len(transcript_content.split()))

        if not transcript_content.strip():
            raise AgentProcessingError("Cannot summarize empty transcript")

        # Create the summarization prompt
        prompt = self._build_summarization_prompt(transcript_content)
        
        # Create system prompt for the summarizer
        system_prompt = self._build_system_prompt()

        try:
            # Get summary from Claude
            raw_summary = self._call_claude(prompt, system_prompt)
            
            # Clean and validate the summary
            summary = self._clean_summary(raw_summary)
            
            # Validate word count
            word_count = len(summary.split())
            if word_count < settings.summary_min_words or word_count > settings.summary_max_words:
                logger.warning(f"[{self.agent_name}] Summary word count outside target range",
                              word_count=word_count,
                              min_words=settings.summary_min_words,
                              max_words=settings.summary_max_words)

            logger.info(f"[{self.agent_name}] Generated summary successfully",
                       summary_word_count=word_count,
                       summary_preview=self._truncate_for_log(summary, 100))

            return {"summary": summary}

        except Exception as e:
            logger.error(f"[{self.agent_name}] Failed to generate summary", error=str(e))
            raise AgentProcessingError(f"Summarization failed: {str(e)}")

    def _build_system_prompt(self) -> str:
        """Build the system prompt for Claude."""
        return f"""You are an expert at creating concise, professional summaries of podcast content for business audiences.

Your task is to create a summary that:
- Is between {settings.summary_min_words} and {settings.summary_max_words} words
- Captures the main topics and themes discussed
- Highlights key insights and important points
- Is written in a professional, business-appropriate tone
- Focuses on factual content rather than opinions
- Does not include filler words or transcription artifacts

The summary should be useful for someone who wants to quickly understand what the podcast covered without listening to it."""

    def _build_summarization_prompt(self, transcript_content: str) -> str:
        """
        Build the prompt for summarizing the transcript.
        
        Args:
            transcript_content: The transcript to summarize
            
        Returns:
            Formatted prompt string
        """
        # Truncate very long transcripts for the prompt
        max_transcript_length = 15000  # Reasonable limit for Claude context
        if len(transcript_content) > max_transcript_length:
            transcript_content = transcript_content[:max_transcript_length] + "\n[...transcript truncated...]"
            logger.info(f"[{self.agent_name}] Truncated long transcript for processing")

        return f"""Please create a professional summary of the following podcast transcript.

The summary should be {settings.summary_min_words}-{settings.summary_max_words} words and cover:
- Main topics and themes discussed
- Key insights and important points made by speakers
- Notable facts or information shared
- Overall context and purpose of the discussion

Focus on the substantive content and avoid including filler words, "ums", or transcription artifacts.

TRANSCRIPT:
{transcript_content}

SUMMARY:"""

    def _clean_summary(self, raw_summary: str) -> str:
        """
        Clean and format the generated summary.
        
        Args:
            raw_summary: Raw summary text from Claude
            
        Returns:
            Cleaned summary text
        """
        # Remove any leading/trailing whitespace
        summary = raw_summary.strip()
        
        # Remove common prefixes that might be added
        prefixes_to_remove = [
            "Summary:",
            "SUMMARY:",
            "Podcast Summary:",
            "This podcast discusses",
            "In this podcast",
            "The podcast covers"
        ]
        
        for prefix in prefixes_to_remove:
            if summary.startswith(prefix):
                summary = summary[len(prefix):].strip()
                break
        
        # Ensure it starts with a capital letter
        if summary and not summary[0].isupper():
            summary = summary[0].upper() + summary[1:]
        
        # Remove extra whitespace and normalize spacing
        summary = re.sub(r'\s+', ' ', summary)
        
        return summary

File: internal/agents/summarizer.go
package agents

import (
	"context"
	"fmt"
	"regexp"
	"strings"
	"time"
	
	"podcast-analyzer/internal/clients"
	"podcast-analyzer/internal/config"
)

// SummarizerAgent generates concise summaries of podcast transcripts
type SummarizerAgent struct {
	*BaseAgent
	anthropicClient clients.AnthropicClientInterface
	maxChars        int
}

// NewSummarizerAgent creates a new summarizer agent
func NewSummarizerAgent(cfg *config.Config) *SummarizerAgent {
	return &SummarizerAgent{
		BaseAgent:       NewBaseAgent("summarizer"),
		anthropicClient: clients.NewAnthropicClient(cfg),
		maxChars:        cfg.SummaryMaxChars,
	}
}

// Process generates a summary of the podcast transcript
func (s *SummarizerAgent) Process(ctx context.Context, content string) (Result, error) {
	start := time.Now()
	defer func() {
		s.LogAPICall(ctx, "anthropic", len(content), true)
	}()
	
	// Log start of processing
	s.LogStart(ctx, len(content))
	
	// Validate content
	if err := s.ValidateContent(content); err != nil {
		s.LogError(ctx, err, time.Since(start))
		return Result{}, err
	}
	
	// Build prompts
	systemPrompt := s.buildSystemPrompt()
	userPrompt := s.buildUserPrompt(content)
	
	// Call Claude API
	rawSummary, err := s.anthropicClient.CallClaude(ctx, s.Name(), userPrompt, systemPrompt, false)
	if err != nil {
		s.LogError(ctx, err, time.Since(start))
		return Result{}, NewAgentError(s.Name(), "failed to generate summary", err)
	}
	
	// Clean and validate the summary
	summary := s.cleanSummary(rawSummary)
	if err := s.validateSummary(summary); err != nil {
		s.LogError(ctx, err, time.Since(start))
		return Result{}, err
	}
	
	result := Result{Summary: summary}
	
	// Log success
	s.LogSuccess(ctx, &result, time.Since(start))
	
	return result, nil
}

// buildSystemPrompt creates the system prompt for Claude
func (s *SummarizerAgent) buildSystemPrompt() string {
	return fmt.Sprintf(`You are an expert at creating concise, professional summaries of podcast content for business audiences.

Your task is to create a summary that:
- Is a maximum of %d characters
- Captures the main topics and themes discussed
- Focuses on factual content rather than opinions
- Does not include filler words or transcription artifacts

The summary should be useful for someone who wants to post a tweet on X or update their status on Facebook.`, s.maxChars)
}

// buildUserPrompt creates the user prompt with the transcript content
func (s *SummarizerAgent) buildUserPrompt(content string) string {
	// Truncate very long transcripts for the prompt
	maxTranscriptLength := 15000 // Reasonable limit for Claude context
	if len(content) > maxTranscriptLength {
		content = s.TruncateContent(content, maxTranscriptLength)
	}
	
	return fmt.Sprintf(`Please create a professional summary of the following podcast transcript.

The summary should be a maximum of %d characters and should include:
- Main topics and themes discussed
- Overall context and purpose of the discussion

TRANSCRIPT:
%s

SUMMARY:`, s.maxChars, content)
}

// cleanSummary cleans and formats the generated summary
func (s *SummarizerAgent) cleanSummary(rawSummary string) string {
	// Remove any leading/trailing whitespace
	summary := strings.TrimSpace(rawSummary)
	
	// Remove common prefixes that might be added by Claude
	prefixes := []string{
		"Summary:",
		"SUMMARY:",
		"Podcast Summary:",
		"This podcast discusses",
		"In this podcast",
		"The podcast covers",
	}
	
	for _, prefix := range prefixes {
		if strings.HasPrefix(summary, prefix) {
			summary = strings.TrimSpace(summary[len(prefix):])
			break
		}
	}
	
	// Ensure it starts with a capital letter
	if len(summary) > 0 && !s.IsUpperCase(summary[0]) {
		summary = strings.ToUpper(string(summary[0])) + summary[1:]
	}
	
	// Remove extra whitespace and normalize spacing
	summary = regexp.MustCompile(`\s+`).ReplaceAllString(summary, " ")
	
	// Ensure it ends with proper punctuation
	if len(summary) > 0 && !strings.HasSuffix(summary, ".") && !strings.HasSuffix(summary, "!") && !strings.HasSuffix(summary, "?") {
		summary += "."
	}
	
	return summary
}

// validateSummary validates the generated summary
func (s *SummarizerAgent) validateSummary(summary string) error {
	if summary == "" {
		return NewAgentError(s.Name(), "generated summary is empty", nil)
	}
	
	if len(summary) > s.maxChars {
		// Log warning but don't fail - truncate if necessary
		s.logger.WithFields(map[string]interface{}{
			"agent":        s.Name(),
			"summary_length": len(summary),
			"max_chars":    s.maxChars,
		}).Warn("Summary exceeds maximum character limit, truncating")
		
		// Truncate to max chars, trying to end at word boundary
		if len(summary) > s.maxChars {
			truncated := summary[:s.maxChars]
			if lastSpace := strings.LastIndex(truncated, " "); lastSpace > s.maxChars-20 {
				truncated = truncated[:lastSpace]
			}
			summary = truncated + "..."
		}
	}
	
	// Check minimum length (very short summaries are probably not useful)
	if len(summary) < 20 {
		return NewAgentError(s.Name(), "summary too short to be meaningful", nil)
	}
	
	return nil
}